habitat-baselines/habitat_baselines/rl/multi_agent/bdp_access_mgr.py

```diff
@@ -0,0 +1,222 @@
+            env_spec.action_space,
+            orig_action_space=env_spec.orig_action_space,
+            agent=self._agents[0],
+            n_agents=num_active_agents,
+            update_obs_with_agent_prefix_fn=self._inject_behav_latent,
+            discrim_reward_weight=self._pop_config.discrim_reward_weight,
+            hl_policy=hl_policy,
+            discrim=discrim,
+        )
+        return multi_policy, multi_updater, multi_storage
+
+    def _create_single_agent(
+        self,
+        config,
+        agent_env_spec,
+        is_distrib,
+        device,
+        use_resume_state,
+        num_envs,
+        percent_done_fn,
+        lr_schedule_fn,
+        agent_name,
+    ):
+        if agent_name == BEHAV_AGENT_NAME:
+            # Inject the behavior latent into the observation spec
+            agent_env_spec.observation_space[BEHAV_ID] = spaces.Box(
+                low=np.finfo(np.float32).min,
+                high=np.finfo(np.float32).max,
+                shape=(self._pop_config.behavior_latent_dim,),
+                dtype=np.float32,
+            )
+        elif agent_name == COORD_AGENT_NAME:
+            # Remove the discriminator from this policy.
+            config = config.copy()
+            with read_write(config):
+                del config.habitat_baselines.rl.auxiliary_losses["bdp_discrim"]
+        else:
+            raise ValueError(f"Unexpected agent name {agent_name}")
+
+        return SingleAgentAccessMgr(
+            config,
+            agent_env_spec,
+            is_distrib,
+            device,
+            use_resume_state,
+            num_envs,
+            percent_done_fn,
+            lr_schedule_fn,
+            agent_name,
+        )
+
+
+@baseline_registry.register_auxiliary_loss(name="bdp_discrim")
+class BehavDiscrim(nn.Module):
+    """
+    Defines the discriminator network and the training objective for the
+    discriminator. Through the Habitat Baselines auxiliary loss registry, this
+    is automatically added to the policy class and the loss is computed in the
+    policy update.
+    """
+
+    def __init__(
+        self,
+        action_space: spaces.Box,
+        net: Net,
+        loss_scale,
+        hidden_size,
+        behavior_latent_dim,
+    ):
+        super().__init__()
+
+        # For now the discriminator input is just (z_t, a_t) where z_t is the
+        # LSTM encoded state and a_t is the current action.
+        input_dim = net._hidden_size
+        self.discrim = nn.Sequential(
+            nn.Linear(input_dim, hidden_size),
+            nn.ReLU(True),
+            nn.Linear(hidden_size, hidden_size),
+            nn.ReLU(True),
+            nn.Linear(hidden_size, behavior_latent_dim),
+        )
+        self.loss_scale = loss_scale
+
+    def pred_logits(self, policy_features, obs):
+        return self.discrim(policy_features)
+
+    def forward(self, policy_features, obs):
+        pred_logits = self.pred_logits(policy_features, obs)
```

Comment:
can't pass obs, since it contains the behavior latent. Remove behavior latent before passing to pred_logits.

Reply:
We actually need to keep it if we decide to generate policy actions in the future (since the policy is latent conditioned).

Reply:
The discriminator cannot have the latent variable as input. Otherwise it will ignore all inputs and just use the latent variable to discriminate the different policies? Am I misunderstanding this somehow?